tests/usm/usm_api.h

```diff
@@ -0,0 +1,688 @@
+using memcpy_from_non_usm = detail::memcpy<T, count, allocation::non_usm>;
+template <typename T, size_t count>
+using memcpy_from_host = detail::memcpy<T, count, allocation::host>;
+template <typename T, size_t count>
+using memcpy_from_device = detail::memcpy<T, count, allocation::device>;
+template <typename T, size_t count>
+using memcpy_from_shared = detail::memcpy<T, count, allocation::shared>;
+
+/** @brief Provides test logic for the fill() method tests
+ */
+template <typename T, size_t count>
+class fill {
+  const T value;
+
+ public:
+  fill(sycl::queue &, int seed) : value{seed} {}
+
+  /** @brief The fill() method doesn't support non-USM pointers
+   */
+  static constexpr bool has_non_usm_support() { return false; }
+
+  template <allocation alloc>
+  static std::string description() {
+    return "fill using " + get_allocation_decription<alloc>();
+  }
+
+  template <typename parentT, typename... depEventsT>
+  auto call(parentT &parent, T *ptr, depEventsT &&... events) const {
+    return parent.fill(ptr, value, count, std::forward<depEventsT>(events)...);
+  }
+
+  void check(const T *ptr) const {
+    T reference[count];
+    std::fill_n(reference, count, value);
+
+    detail::check_values<count>(ptr, reference);
+  }
+};
+
+/** @brief Provides test logic for the memset() method tests
+ */
+template <typename T, size_t count>
+class memset {
+  const int value;
+
+ public:
+  static constexpr size_t size = count * sizeof(T);
+
+  memset(sycl::queue &, int seed) : value(seed) {}
+
+  /** @brief The memset() method doesn't support non-USM pointers
+   */
+  static constexpr bool has_non_usm_support() { return false; }
+
+  template <allocation alloc>
+  static std::string description() {
+    return "memset using " + get_allocation_decription<alloc>();
+  }
+
+  template <typename parentT, typename... depEventsT>
+  auto call(parentT &parent, T *ptr, depEventsT &&... events) const {
+    return parent.memset(ptr, value, size, std::forward<depEventsT>(events)...);
+  }
+
+  void check(const T *ptr) const {
+    T reference[count];
+    std::memset(reference, value, size);
+
+    detail::check_values<count>(ptr, reference);
+  }
+};
+
+/** @brief Provides test logic for the prefetch() method tests
+ */
+template <typename T, size_t count>
+class prefetch {
+ public:
+  static constexpr size_t size = count * sizeof(T);
+
+  prefetch(sycl::queue &, int) {}
+
+  /** @brief The prefetch() method doesn't support non-USM pointers
+   */
+  static constexpr bool has_non_usm_support() { return false; }
+
+  template <allocation alloc>
+  static std::string description() {
+    return "prefetch using " + get_allocation_decription<alloc>();
+  }
+
+  template <typename parentT, typename... depEventsT>
+  auto call(parentT &parent, T *ptr, depEventsT &&... events) const {
+    return parent.prefetch(ptr, size, std::forward<depEventsT>(events)...);
+  }
+
+  void check(const T *) const {
+    // Results are implementation-defined
```

Comment:
I hope not. What about something like
```suggestion
    // Result is not really observable
```
?

Reply:
We mean that result is an implementation-defined [SYCL_Docs](https://github.com/KhronosGroup/SYCL-Docs/blob/SYCL-2020/master/adoc/chapters/programming_interface.adoc#:~:text=This%20operation%20is%20an%20implementation-defined) about this

Maybe we can replace "Results" to some another word?